Fix crash previously fixed in d14ed91. Fixes #2270

bartekfi opened the pull request:
Closes #2270 

#### What has been done to verify that this works as intended?

I was not able to reproduce it. The only verification is the logical analysis of the code. I will test the solution once I understand the steps @grzesiek2010 performed to fix https://github.com/opendatakit/collect/issues/998

#### Why is this the best possible solution? Were any other approaches considered?

This is an unverified fix based on assumptions from changes made first in:
https://github.com/opendatakit/collect/issues/998 to fix an issue, 
and then in:
https://github.com/opendatakit/collect/pull/1038
that probably caused the issue to reemerge. 

The changes in #998 introduced a null check to make sure the `FormHierarchyActivity` is closed when FormController is null. 

#1038 created a subclass of `FormHierarchyActivity` -> `ViewFormHierarchyActivity`. This caused a reintroduction of the bug fixed in #998, as the subclass does not recheck for null in the onCreate() method.

When the same situation that caused crash fixed in #998 occurs in the ViewFormHierarchyActivity following steps will happen:

1. ViewFormHierarchyActivity will start with calling [super.onCreate()](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/ViewFormHierarchyActivity.java#L38)
2. A null check in `super.onCreate()` parent class [FormHierarchyActivity::onCreate ](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java#L81) will call shortcut the `onCreate()` method by calling return, and call `finish()` on activity. 
3. `finish()` will be scheduled to close the activity asynchronously after the `onCreate()` is finished
4. `super.onCreate()` is done, the rest of the `onCreate()` method in ViewFormHierarchyActivity will be called. The shortcut only affected stopping `super.onCraete()`, the `onCreate()` is still called.
5. `Collect.getInstance().getFormController()` will still be null. It needs to be to be shortcut again. No need to call finish() this time, as it was already called in the `super.onCreate()`.


#### Are there any risks to merging this code? If so, what are they?

No changes to business logic. The activity silently finishes instead of throwing a Null Pointer Exception. 

#### Do we need any specific form for testing your changes? If so, please attach one.

The issue should be reproduced by the same steps as described in https://github.com/opendatakit/collect/issues/998

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lint findbugs` and confirmed all checks still pass.

bartekfi commented:
@grzesiek2010 Can you describe the steps you took to fix issue #998 - I think the two cases the same bug but reintroduced.

grzesiek2010 commented:
It's not even the same bug in those two cases but in many different cases. I started an investigation here https://github.com/opendatakit/collect/issues/1533

When it comes to #998 I described steps I used to reproduce the problem.

> Got it! It takes place when we close the app (when the FormHierarchyActivity is opened) using a physical button and leave the app for a long time (in my case it's at least 30 minutes), then if we open the app again the app crashes.

Unfortunately, as I said the problem occurs under various different circumstances and only sometimes we are able to reproduce it, so the only solution, for now, is just checking if our `FormController` object is null.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/ViewFormHierarchyActivity.java:
```diff
@@ -37,7 +38,11 @@
-        Collect.getInstance().getFormController().stepToOuterScreenEvent();
+        FormController formController = Collect.getInstance().getFormController();
+        if (formController == null) {
+            return;
```
I don't think `return` is enough here, that means you just stop onCreate() method but the activity is still displayed, probably we should finish in such a case.

bartekfi replied:
I see how this is confusing. Finish is actually called in the super class based on the same condition. 

What I can do instead, is check here if the activity is actually finishing, by calling `isFinished()`. This will reduce code duplication and rely on the finishing condition.

grzesiek2010 replied:
Ohh I get it so `return` seems fine.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/ViewFormHierarchyActivity.java:
```diff
@@ -37,7 +38,11 @@
+        FormController formController = Collect.getInstance().getFormController();
+        if (formController == null) {
+            return;
+        }
```
please add `else` clause and log the issue, the same reason like here https://github.com/opendatakit/collect/pull/2277#issuecomment-395383314

bartekfi replied:
Makes sense. I'll add logging in the parent class, as this should be the single point of checking for this error. Thanks for pointing that out.

Commit:
Don't proceed with onCreate if the activity is already finishing

Parent activity can call finish() if it detects that FormController is
null. In that case ViewFormHierarchyActivity should respect this
decision to end its own life, and not interfere. (that's grim)

Commit:
Log when FormController is null

bartekfi commented:
@grzesiek2010 I've updated the flow to look at the finish instead of FormController. I think this way we don't duplicate the same check, and expose the mechanism that caused this issue to reemerge - that is not respecting finish() called in parent class. I've also added logging.

lognaturel commented:
Thank you, @bartoszfilipowicz! I like your grim commit comment. 💀 Going to give @grzesiek2010 a chance to review.

grzesiek2010 commented:
It looks good.
